File: src/lib/missions/handlers/TransportMission.ts
/**
 * TransportMission Handler
 *
 * Handles transport missions that deliver resources to another planet.
 * - On arrival: delivers resources to target planet, fleet returns empty
 * - On return: fleet returns to origin planet
 *
 * Transport missions can be sent to any planet (own or other players')
 */

import type { MissionType } from '@/types/database'
import { BaseMission } from '../BaseMission'
import {
  MissionContext,
  MissionArrivalResult,
  MissionReturnResult,
  Resources,
  emptyResources,
} from '../types'

export class TransportMission extends BaseMission {
  readonly missionType: MissionType = 'transport'
  readonly hasReturn: boolean = true
  readonly name: string = 'Transport'

  /**
   * Process transport mission arrival
   *
   * 1. Deliver resources to target planet
   * 2. Send messages to both players (if different)
   * 3. Fleet returns empty
   */
  async processArrival(context: MissionContext): Promise<MissionArrivalResult> {
    const { mission, originPlanet, targetPlanet } = context

    // Target planet must exist for transport
    if (!targetPlanet) {
      return this.errorArrival('Target planet not found')
    }

    const resources = this.getResources(mission)
    const ships = this.getShips(mission)

    // Add resources to target planet
    const addResult = await this.addResourcesToPlanet(targetPlanet.id, resources)
    if (!addResult.success) {
      return this.errorArrival(`Failed to deliver resources: ${addResult.error}`)
    }

    // Format coordinates for messages
    const originCoords = this.formatCoords(
      mission.origin_galaxy,
      mission.origin_system,
      mission.origin_position
    )
    const targetCoords = this.formatCoords(
      mission.destination_galaxy,
      mission.destination_system,
      mission.destination_position
    )

    // Create messages
    const messages = []

    // Message to fleet owner (origin player)
    const ownerMessage = this.createMessage(
      mission.user_id,
      'transport',
      'Transport Arrived',
      `Your transport fleet has arrived at ${targetCoords} and delivered:\n` +
        `Metal: ${resources.metal.toLocaleString()}\n` +
        `Crystal: ${resources.crystal.toLocaleString()}\n` +
        `Deuterium: ${resources.deuterium.toLocaleString()}\n` +
        `The fleet is now returning to ${originCoords}.`
    )
    messages.push(ownerMessage)

    // If target is a different player, send them a message too
    if (targetPlanet.user_id !== mission.user_id) {
      const receiverMessage = this.createMessage(
        targetPlanet.user_id,
        'transport',
        'Transport Received',
        `A transport fleet from ${originCoords} has delivered resources to your planet ${targetCoords}:\n` +
          `Metal: ${resources.metal.toLocaleString()}\n` +
          `Crystal: ${resources.crystal.toLocaleString()}\n` +
          `Deuterium: ${resources.deuterium.toLocaleString()}`
      )
      messages.push(receiverMessage)
    }

    // Return with ships but no resources (all delivered)
    return this.successArrival(
      true, // shouldReturn
      emptyResources(), // returnResources - empty, all delivered
      ships, // returnShips - all ships return
      messages,
      []
    )
  }

  /**
   * Process transport mission return
   *
   * 1. Add ships back to origin planet
   * 2. Send return message to player
   */
  async processReturn(context: MissionContext): Promise<MissionReturnResult> {
    const { mission, originPlanet } = context

    if (!originPlanet) {
      return this.errorReturn('Origin planet not found')
    }

    const ships = this.getShips(mission)
    const resources = this.getResources(mission)

    // Add ships back to origin planet
    const addShipsResult = await this.addShipsToPlanet(originPlanet.id, ships)
    if (!addShipsResult.success) {
      return this.errorReturn(`Failed to return ships: ${addShipsResult.error}`)
    }

    // If there are any resources (shouldn't happen in normal transport, but handle it)
    if (resources.metal > 0 || resources.crystal > 0 || resources.deuterium > 0) {
      const addResourcesResult = await this.addResourcesToPlanet(originPlanet.id, resources)
      if (!addResourcesResult.success) {
        return this.errorReturn(`Failed to return resources: ${addResourcesResult.error}`)
      }
    }

    // Format coordinates
    const originCoords = this.formatCoords(
      mission.destination_galaxy, // For return, destination is the original origin
      mission.destination_system,
      mission.destination_position
    )
    const returnFromCoords = this.formatCoords(
      mission.origin_galaxy, // For return, origin is where we're coming from
      mission.origin_system,
      mission.origin_position
    )

    // Create return message
    const message = this.createMessage(
      mission.user_id,
      'transport',
      'Fleet Returned',
      `Your fleet has returned from transport mission to ${returnFromCoords} and arrived at ${originCoords}.`
    )

    return this.successReturn([message], [])
  }
}

File: unity_client/Assets/Scripts/SharedLib/Missions/Handlers/TransportMission.cs
// =============================================================================
// TransportMission.cs — Transport mission handler
// Port of: src/lib/missions/handlers/TransportMission.ts
// Namespace: OGameX.SharedLib.Missions.Handlers
// =============================================================================

using System.Collections.Generic;

namespace OGameX.SharedLib.Missions.Handlers
{
    /// <summary>
    /// Handles transport missions that deliver resources to another planet.
    /// - On arrival: delivers resources to target planet, fleet returns empty
    /// - On return: fleet returns to origin planet
    /// Transport missions can be sent to any planet (own or other players').
    /// </summary>
    public class TransportMission : BaseMission
    {
        public override MissionType HandlerMissionType => MissionType.Transport;
        public override bool HasReturn => true;
        public override string Name => "Transport";

        /// <summary>
        /// Process transport mission arrival.
        /// 1. Deliver resources to target planet
        /// 2. Send messages to both players (if different)
        /// 3. Fleet returns empty
        /// </summary>
        public override MissionArrivalResult ProcessArrival(MissionContext context)
        {
            var mission = context.Mission;
            var targetPlanet = context.TargetPlanet;

            // Target planet must exist
            if (targetPlanet == null)
                return ErrorArrival("Target planet not found");

            var resources = GetResources(mission);
            var ships = GetShips(mission);

            // Format coordinates
            var originCoords = FormatCoords(
                mission.OriginGalaxy, mission.OriginSystem, mission.OriginPosition);
            var targetCoords = FormatCoords(
                mission.DestinationGalaxy, mission.DestinationSystem, mission.DestinationPosition);

            // Create messages
            var messages = new List<MissionMessage>();

            // Message to fleet owner
            messages.Add(CreateMessage(
                mission.UserId,
                MessageType.Transport,
                "Transport Arrived",
                $"Your transport fleet has arrived at {targetCoords} and delivered:\n" +
                FormatResources(resources) + "\n" +
                $"The fleet is now returning to {originCoords}."
            ));

            // If target is a different player, notify them too
            if (targetPlanet.UserId != mission.UserId)
            {
                messages.Add(CreateMessage(
                    targetPlanet.UserId,
                    MessageType.Transport,
                    "Transport Received",
                    $"A transport fleet from {originCoords} has delivered resources to your planet {targetCoords}:\n" +
                    FormatResources(resources)
                ));
            }

            // Return with ships but no resources (all delivered)
            return SuccessArrival(
                shouldReturn: true,
                returnResources: Resources.Empty(),
                returnShips: ships,
                messages: messages
            );
        }

        /// <summary>
        /// Process transport mission return.
        /// 1. Add ships back to origin planet
        /// 2. Send return message
        /// </summary>
        public override MissionReturnResult ProcessReturn(MissionContext context)
        {
            var mission = context.Mission;
            var originPlanet = context.OriginPlanet;

            if (originPlanet == null)
                return ErrorReturn("Origin planet not found");

            var originCoords = FormatCoords(
                mission.DestinationGalaxy, mission.DestinationSystem, mission.DestinationPosition);
            var returnFromCoords = FormatCoords(
                mission.OriginGalaxy, mission.OriginSystem, mission.OriginPosition);

            var message = CreateMessage(
                mission.UserId,
                MessageType.Transport,
                "Fleet Returned",
                $"Your fleet has returned from transport mission to {returnFromCoords} and arrived at {originCoords}."
            );

            return SuccessReturn(new List<MissionMessage> { message });
        }
    }
}
